src/read.rs

```diff
@@ -205,3 +206,148 @@ fn read_exact_eof<R: Read>(rdr: &mut R, buf: &mut [u8]) -> io::Result<bool> {
+    src: Vec<u8>,
+
+    /// Have we written the standard snappy header to `dst` yet?
+    wrote_stream_ident: bool,
+}
+
+impl<R: Read> FrameEncoder<R> {
+    /// Create a new reader for streaming Snappy compression.
+    pub fn new(rdr: R) -> FrameEncoder<R> {
+        FrameEncoder {
+            inner: Inner {
+                r: rdr,
+                enc: Encoder::new(),
+                src: Vec::with_capacity(MAX_BLOCK_SIZE),
+                wrote_stream_ident: false,
+            },
+            dst: vec![0; *MAX_READ_FRAME_ENCODER_BLOCK_SIZE],
+            dsts: 0,
+            dste: 0,
+        }
+    }
+
+    /// Gets a reference to the underlying reader in this decoder.
+    pub fn get_ref(&self) -> &R {
+        &self.inner.r
+    }
+
+    /// Read previously compressed data from `self.dst`, returning the number of
+    /// bytes read. If `self.dst` is empty, returns 0.
+    fn read_from_dst(&mut self, buf: &mut [u8]) -> usize {
+        let available_bytes = self.dste - self.dsts;
+        let count = cmp::min(available_bytes, buf.len());
+        buf[..count].copy_from_slice(&self.dst[self.dsts..self.dsts+count]);
+        self.dsts += count;
+        count
+    }
+}
+
+impl<R: Read> Read for FrameEncoder<R> {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        // Try reading previously compressed bytes from our `dst` buffer, if
+        // any.
+        let count = self.read_from_dst(buf);
+
+        if count > 0 {
+            // We had some bytes in our `dst` buffer that we used.
+            Ok(count)
+        } else if buf.len() >= *MAX_READ_FRAME_ENCODER_BLOCK_SIZE {
+            // Our output `buf` is big enough that we can directly write into
+            // it, so bypass `dst` entirely.
+            self.inner.read_frame(buf)
+        } else {
+            // We need to refill `self.dst`, and then return some bytes from
+            // that.
+            let count = try!(self.inner.read_frame(&mut self.dst));
+            self.dsts = 0;
+            self.dste = count;
+            Ok(self.read_from_dst(buf))
+        }
+    }
+}
+
+impl<R: Read> Inner<R> {
+    /// Read from `self.r`, and create a new frame, writing it to `dst`, which
+    /// must be at least `*MAX_READ_FRAME_ENCODER_BLOCK_SIZE` bytes in size.
+    fn read_frame(&mut self, dst: &mut [u8]) -> io::Result<usize> {
+        debug_assert!(dst.len() >= *MAX_READ_FRAME_ENCODER_BLOCK_SIZE);
+
+        // Try to read a max-sized block to compress from the underlying stream.
+        // This is surprisingly complicated in Rust, requiring us to pass a
+        // zero-length mutable buffer and get it filled, and requiring the
+        // use of the `by_ref().take(...).read_to_end(...)` idiom to read
+        // up to the specified number of bytes from a file.
+        self.src.truncate(0);
+        try!(self.r.by_ref().take(MAX_BLOCK_SIZE as u64)
+            .read_to_end(&mut self.src));
```

Comment:
I know it's been a while, but do you remember what led you to do this instead of just `self.r.read(&mut self.src)`? Granted, the latter is somewhat at the mercy of the underlying reader. If the underlying reader emits very small reads, then compression is going to be pretty stinky. On the other hand, this approach means that the reader will stubbornly wait until it has a pretty beefy number of bytes or until it hits EOF. I guess I'm just trying to picture what happens when the reader is a blocking network socket. Should our reader really try to sit and wait until its buffer is full?

I'm not a compression expert, so I suppose we should probably do whatever others do... So I drilled down into the `flate2` crate, and it _looks_ like it does indeed just issue a `read` call without trying to fill up as many bytes as it can: https://github.com/alexcrichton/flate2-rs/blob/4d62a8936981a74a377459155e6d41333c66cf62/src/zio.rs#L118-L157 (There is a loop, but that loop is just trying to make sure it reads a non-zero number of bytes.)

Thinking about this more, I think `flate2`'s behavior is "better." While either choice can result in bad things, I think 1) potentially blocking like this is perhaps more surprising and 2) since `self.src` is appropriately sized, a "well behaved" implementation of `std::io::Read` really should fill up the buffer with as many bytes as it can. For sure, OS level `read` calls generally do this.